### PR TITLE
SONARPY-1335 Rule S6553: "null=True": should not be used on string-based fields in Django models

### PR DESCRIPTION
**Pull request by maksim-grebeniuk-sonarsource**

### CODE REVIEW

File: python-checks/src/test/resources/checks/django/djangoModelStringFieldCheck.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+class NullBlankFieldsModel(models.Model):
+    name = models.CharField(max_length=50, null=True, blank=True) # Noncompliant {{Remove this "null=True" flag.}}
+    desc = models.TextField(max_length=50, null=True, blank=True) # Noncompliant
+
+class BlankFieldsModel(models.Model):
+    name = models.CharField(max_length=50, blank=True)
+    desc = models.TextField(max_length=50, blank=True)
+
+
```
Comment (guillaume-dequenne-sonarsource) [inline python-checks/src/test/resources/checks/django/djangoModelStringFieldCheck.py:16]: One case that seems to be missing is when both `unique=True` and `blank=True`, where we should not raise an issue.

File: python-checks/src/main/java/org/sonar/python/checks/django/DjangoModelStringFieldCheck.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+      .filter(FIELD_TYPES_FQN::contains)
+      .isPresent();
+  }
+
+  private static void validateTextFieldArguments(SubscriptionContext ctx, CallExpression call) {
+    var nullArg = getCallArgumentByName(call, "null")
+      .filter(DjangoModelStringFieldCheck::isArgumentSetAsTrue);
+    var blankArg = getCallArgumentByName(call, "blank")
+      .filter(DjangoModelStringFieldCheck::isArgumentSetAsTrue);
+
```
Comment (guillaume-dequenne-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/django/DjangoModelStringFieldCheck.java:83]: We can probably look for `unique=True` here and if it's present as well as `blankArg`, we can abort.

File: python-checks/src/main/java/org/sonar/python/checks/django/DjangoModelStringFieldCheck.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  private static Optional<PythonQuickFix> createRemoveArgQuickFix(CallExpression call, Tree arg) {
+    return Optional.ofNullable(call.argumentList())
+      .map(Tree::children)
+      .map(args -> {
+        var count = args.size();
+        var index = args.indexOf(arg);
+        var removeFrom = arg;
+        Tree removeTo;
+        if (index == 0) {
```
Comment (guillaume-dequenne-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/django/DjangoModelStringFieldCheck.java:128]: I don't like too much when we do this kind of index arithmetic. If we can't simplify, maybe it's good to have at least some comments that clarify what we're doing in each branch.

**Review by guillaume-dequenne-sonarsource — CHANGES_REQUESTED** [review]
Looking good.
There is one exception case that isn't covered that we should probably cover, though.
I left some other minor suggestions, nothing big though.

File: python-checks/src/test/java/org/sonar/python/checks/django/DjangoModelStringFieldCheckTest.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+    var after = "from django.db import models\n" +
+      "class NullFieldsModel(models.Model):\n" +
+      "    name = models.CharField(max_length=50, blank=True)";
+    PythonQuickFixVerifier.verify(check, before, after);
+    PythonQuickFixVerifier.verifyQuickFixMessages(check, before, "Replace with \"blank=True\"");
+  }
+
+  @Test
+  public void removeQuickFixTest() {
```
[the remark this answers is not in the document]
Comment (maksim-grebeniuk-sonarsource) [reply python-checks/src/test/java/org/sonar/python/checks/django/DjangoModelStringFieldCheckTest.java:48]: by our tests I can see that it removes the line

File: python-checks/src/main/java/org/sonar/python/checks/django/DjangoModelStringFieldCheck.java (hunk 3 above)
Comment (maksim-grebeniuk-sonarsource) [reply python-checks/src/main/java/org/sonar/python/checks/django/DjangoModelStringFieldCheck.java:134]: done

File: python-checks/src/main/java/org/sonar/python/checks/django/DjangoModelStringFieldCheck.java (hunk 2 above)
Comment (maksim-grebeniuk-sonarsource) [reply python-checks/src/main/java/org/sonar/python/checks/django/DjangoModelStringFieldCheck.java:89]: done

File: python-checks/src/test/resources/checks/django/djangoModelStringFieldCheck.py (hunk 1 above)
Comment (maksim-grebeniuk-sonarsource) [reply python-checks/src/test/resources/checks/django/djangoModelStringFieldCheck.py]: done

File: python-checks/src/main/java/org/sonar/python/checks/django/DjangoModelStringFieldCheck.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+import org.sonar.python.quickfix.TextEditUtils;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S6553")
+public class DjangoModelStringFieldCheck extends PythonSubscriptionCheck {
+
+  private static final String REPLACE_MESSAGE = "Replace this \"null=True\" flag with \"blank=True\".";
+  private static final String REMOVE_MESSAGE = "Remove this \"null=True\" flag.";
+  private static final String REPLACE_QUICK_FIX_MESSAGE = "Replace with \"blank=True\"";
+  private static final String REMOVE_QUICK_FIX_MESSAGE = "Remove \"null\" argument";
```
[the remark this answers is not in the document]
Comment (maksim-grebeniuk-sonarsource) [reply python-checks/src/main/java/org/sonar/python/checks/django/DjangoModelStringFieldCheck.java]: done

File: python-checks/src/main/java/org/sonar/python/checks/django/DjangoModelStringFieldCheck.java (hunk 3 above)
Comment (guillaume-dequenne-sonarsource) [reply python-checks/src/main/java/org/sonar/python/checks/django/DjangoModelStringFieldCheck.java:134]: Thanks!

**Review by guillaume-dequenne-sonarsource — APPROVED** [review]
LGTM!
It seems you need to rebase before merging.

**Pull request merged**